Add multiple face group merge support

marc7s opened the pull request:
Reopening this PR now that I have migrated the code to the new photoview `master`. The old PR is #912.

From the previous PR there are some changes:
1. Various UI improvements (some selection highlighting, cursor types etc to make it more user friendly)
2. Improved backend code without loops for merging multiple groups
3. Tracked down the issue I could not resolve with duplicates entries

The most important of these was number 3, with the bug I could not find a pattern for reproducing. I have now finally tracked it down and figured out that it was unrelated to my code, so I will open an issue for it. After my previous PR work, my server got wiped and so it has taken me a long while to get everything set up again, finding the time and motivation to do it.

To make this PR more clear, I will copy over relevant information from the previous PR below:

------

Changes the MergeFaceGroupsModal to support merging multiple face groups at the same time.

This is something that I have found very tedious, only being able to merge one face group at a time. With this change, the MergeFaceGroupsModal now has two pages, the first where you select the destination face group, and then the second where you select one or more source face groups. It has also been requested in issue #718.

marc7s commented on ui/src/extractedTranslations/pt/translation.json:
```diff
@@ -64,6 +64,7 @@
     "action_label": {
       "change_label": "",
       "detach_images": "",
+      "merge_face": "",
```
Cannot validate portuguese translation

marc7s commented on api/graphql/resolvers/faces.util.go:
```diff
@@ -94,3 +95,16 @@ func deleteEmptyFaceGroups(sourceFaceGroups []*models.FaceGroup, tx *gorm.DB) er
 	return nil
 }
+
+func deleteFaceGroups(sourceFaceGroups []*models.FaceGroup, tx *gorm.DB) error {
+	for _, faceGroup := range sourceFaceGroups {
+		if err := tx.Model(&models.ImageFace{}).Where(faceGroupIDIsQuestion, faceGroup.ID).Error; err != nil {
+			return err
+		}
+
+		if err := tx.Delete(&faceGroup).Error; err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
Not sure what this is about. This would not really verify anything? The `deleteEmptyFaceGroups` has to check the count as it only deletes the groups if they are empty. It is used when moving faces, since moving faces may potentially leave an empty group behind which should be cleaned up by that function.

But for this function, we already know it will be empty since it was merged, so I removed the check. I could remove this function and call `deleteEmptyFaceGroups` in the merge code as well, but if the code works as expected there is no difference.

Opinions?

marc7s replied:
I do not know enough about golang and these DB calls to judge this so I will leave this open waiting for comments from real people

kkovaletp replied:
@googollee, can you please look into this case: is the suggestion valid?

marc7s commented on ui/src/extractedTranslations/de/translation_old.json:
```diff
@@ -53,7 +53,7 @@
       },
       "merge_face_groups": {
-        "description": null,
+        "description": "Alle Bilder dieser nicht Zugeordneten Person werden mit folgender Person zusammengeführt",
```
This was created by the generator, not me. I will revert the changes as I cannot validate this

marc7s commented on ui/src/extractedTranslations/fr/translation_old.json:
```diff
@@ -43,7 +43,7 @@
       },
       "merge_face_groups": {
-        "description": null,
+        "description": "Toutes les images de ce groupe de visages seront fusionnées dans le groupe de visages sélectionné.",
```
Again, don't know where this came from. I will revert

marc7s commented on ui/src/extractedTranslations/it/translation_old.json:
```diff
@@ -40,6 +40,9 @@
       }
     },
     "modal": {
+      "merge_face_groups": {
+        "description": "Tutte le immagini di questo volto saranno unite alle immagini del volto selezionato"
+      },
```
Same again, I will revert

marc7s commented on api/graphql/generated.go:
```diff
@@ -2081,21 +2081,21 @@ func (ec *executionContext) field_Mutation_combineFaceGroups_argsDestinationFace
-	if _, ok := rawArgs["sourceFaceGroupID"]; !ok {
-		var zeroVal int
+) ([]int, error) {
+	if _, ok := rawArgs["sourceFaceGroupIDs"]; !ok {
+		var zeroVal []int
 		return zeroVal, nil
 	}
 
-	ctx = graphql.WithPathContext(ctx, graphql.NewPathWithField("sourceFaceGroupID"))
-	if tmp, ok := rawArgs["sourceFaceGroupID"]; ok {
-		return ec.unmarshalNID2int(ctx, tmp)
+	ctx = graphql.WithPathContext(ctx, graphql.NewPathWithField("sourceFaceGroupIDs"))
+	if tmp, ok := rawArgs["sourceFaceGroupIDs"]; ok {
+		return ec.unmarshalNID2ᚕintᚄ(ctx, tmp)
 	}
 
-	var zeroVal int
+	var zeroVal []int
```
Commenting on generated code

kkovaletp commented on ui/src/Pages/PeoplePage/SingleFaceGroup/MergeFaceGroupsModal.tsx:
```diff
@@ -1,9 +1,9 @@
 import { gql, PureQueryOptions, useMutation, useQuery } from '@apollo/client'
-import React, { useState } from 'react'
+import React, { useEffect, useState } from 'react'
```
@marc7s, when you mark a thread as resolved, please add a message with at least short description of the resolution: was it implemented and committed, or discarded as invalid, etc.

marc7s replied:
It does not really propose anything here, so nothing to fix. Things like these should really be fixed by a linter and not through PRs, we should have an automatic linter that does things like this

kkovaletp commented:
@marc7s, thanks for your huge work on this feature!
Can you please add short explanations to the resolved threads from the AI? It would help us to understand why you decided to resolve them and how exactly.

@googollee, please review the backend part.

marc7s commented on ui/src/Pages/PeoplePage/SingleFaceGroup/MergeFaceGroupsModal.tsx:
```diff
@@ -59,59 +66,194 @@ const MergeFaceGroupsModal = ({
+      open: isOpen,
+    },
+    searchTitle:
+      t(
+        'people_page.modal.merge_face_groups.sources_table.title',
+        'Select one or more source faces to merge into:'
+      ) +
+      ` ${
+        selectedDestinationFaceGroup?.label ??
+        t('people_page.face_group.unlabeled', 'Unlabeled') ??
+        'Unlabeled'
+      }`,
```
Just formatting stuff. Personally I do not think this is more readable, so I ignored it

marc7s commented on ui/src/Pages/PeoplePage/PeoplePage.tsx:
```diff
@@ -269,31 +276,61 @@ export const PeoplePage = () => {
+        ]}
+      />
+    </>
+  )
```
Fixed in 2b9b443084e368ef20ec0e1ca1a5c4cee1bba7a8

marc7s commented on ui/src/Pages/PeoplePage/PeoplePage.tsx:
```diff
@@ -269,31 +276,59 @@ export const PeoplePage = () => {
+            {t('people_page.action_label.merge_people', 'Merge people')}
+          </Button>
+        </li>
+      </ul>
```
Aria labels added in ce8b0b451f306525977d55eec1866285f851d949
Roles should not be added as per [best practices](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listitem_role#best_practices)

marc7s commented:
> Can you please add short explanations to the resolved threads from the AI?

Sure, I have gone through them now. However, I find the AI mostly annoying as it generates so many comments, many of which are useless for various reasons. For example commenting on generated code, hallucinations or generating one comment for each instance of the same issue. It adds a bunch of additional work for me to go through all the comments to look for the occasional good feedback. Perhaps it is better for other parts of the project, but for the frontend I think an improved developer experience with automated linting and pre-commit checks etc would be more useful than the AI for most things.

marc7s commented:
Any update on this?

marc7s commented:
@kkovaletp My grasp on the technology stack is good enough to do some development, but I don't have a good overall view. Unfortunately even if I did, I don't have enough time at the moment to do code reviews and other work, and I don't know any frontend developers.

kkovaletp commented:
@marc7s, ok, thanks, I see.
Then, I'll keep searching and asking)

marc7s commented:
@googollee Could you review this PR please? Hoping to get this merged

googollee commented:
@marc7s I'll review it this week. Thanks for the contribution.

googollee commented on api/graphql/resolvers/faces.util.go:
```diff
@@ -94,3 +95,16 @@ func deleteEmptyFaceGroups(sourceFaceGroups []*models.FaceGroup, tx *gorm.DB) er
+
+func deleteFaceGroups(sourceFaceGroups []*models.FaceGroup, tx *gorm.DB) error {
+	for _, faceGroup := range sourceFaceGroups {
+		if err := tx.Model(&models.ImageFace{}).Where(faceGroupIDIsQuestion, faceGroup.ID).Error; err != nil {
```
I didn't understand this. `tx.Model().Where()` won't execute any SQL. It just constructs a SQL query.

What do you want to check here? Checking if `facegroup.ID` exists?

marc7s replied:
Not sure what it was meant to do, it was a long time ago this was first written. I tried to find references to why it was added but was unable to, so I removed it in 722591d316b97c20509a3f68cbc4d29dbe20d1df and it still works.

googollee commented on api/graphql/resolvers/faces.util.go:
```diff
@@ -94,3 +95,16 @@ func deleteEmptyFaceGroups(sourceFaceGroups []*models.FaceGroup, tx *gorm.DB) er
+		}
+
+		if err := tx.Delete(&faceGroup).Error; err != nil {
+			return err
```
It's better to add context into the error. Like:

```
return fmt.Errorf("Delete FaceGroup(%d) error: %w", err)
```

It's better to add similar context in line 102, but I can't understand it.

marc7s replied:
Fixed in 722591d316b97c20509a3f68cbc4d29dbe20d1df

googollee reviewed (commented):
Please rebase with the latest code. If I remember correctly, there is a new language zh-TW, which is missing in this PR. Thank you.

googollee reviewed (approved):
Look good to me. I'll update Chinese translation after this PR merged.

kkovaletp commented:
@googollee, did you review the frontend part too?
Are we ok with the current level of test coverage? This looks to be a big feature with many changes... I'd cover such changes  to make sure there are no big issues and it works as expected...

@marc7s, please provide your feedback to the CodeRabbit's review comments and also look onto the comments inside the https://github.com/photoview/photoview/pull/1162#pullrequestreview-2819133384 message

kkovaletp commented:
> I'll update Chinese translation after this PR merged.

@googollee, do you know that you can add your commits to this PR before the merge? Doesn't it look more consistent to make all related changes in the branch before the merge? 
Just asking)

googollee commented:
> @googollee, did you review the frontend part too? Are we ok with the current level of test coverage? This looks to be a big feature with many changes... I'd cover such changes to make sure there are no big issues and it works as expected...

I do check frontend code, but I don't have many experience here. I agree that it should be covered by test.

> @googollee, do you know that you can add your commits to this PR before the merge? Doesn't it look more consistent to make all related changes in the branch before the merge? Just asking)

I guess that means I need to pr to @marc7s branch. I'd like to do that if @marc7s finishes everything.

marc7s commented on ui/src/Pages/PeoplePage/SingleFaceGroup/MergeFaceGroupsModal.tsx:
```diff
@@ -59,59 +66,194 @@
-        title={t(
-          'people_page.modal.merge_face_groups.destination_table.title',
-          'Select the destination face'
-        )}
+        title={modalContent.searchTitle}
+        frozen={state === MergeFaceGroupsModalState.SelectDestination && preselectedDestinationFaceGroup !== undefined}
         faceGroups={filteredFaceGroups}
-        selectedFaceGroup={selectedFaceGroup}
-        setSelectedFaceGroup={setSelectedFaceGroup}
+        selectedFaceGroups={selectedFaceGroups}
+        toggleSelectedFaceGroup={(face) => { if(face === null) return; handleFaceGroupToggled(face) }}
       />
     </Modal>
```
Added tests in a41da89

marc7s commented:
@kkovaletp I have added tests and gone through the AI comments. The only one remaining is https://github.com/photoview/photoview/pull/1162#discussion_r1929630004 where I am hoping @googollee can provide insights.

@googollee I have added tests, rebased from main to fix the conflict and ran reran the code generation. So perhaps the translation task is now fixed also

marc7s commented on ui/src/Pages/PeoplePage/SingleFaceGroup/MergeFaceGroupsModal.tsx:
```diff
@@ -59,59 +66,194 @@
+    if (isNil(selectedDestinationFaceGroup))
+      throw new Error('No selected destination face group')
+
+    const sourceGroupIDs: string[] = [...selectedFaceGroups].filter(fc => fc !== null).map(fc => fc.id)
+
+    if(sourceGroupIDs.length < 1)
+      throw new Error('No selected source face groups')
 
     combineFacesMutation({
       variables: {
-        srcID: sourceFaceGroup.id,
-        destID: selectedFaceGroup.id,
+        srcIDs: sourceGroupIDs,
+        destID: selectedDestinationFaceGroup.id,
       },
     }).then(() => {
-      setOpen(false)
-      navigate(`/people/${selectedFaceGroup.id}`)
+      setState(MergeFaceGroupsModalState.Closed)
+      navigate(`/people/${selectedDestinationFaceGroup.id}`)
     })
   }
```
Every time I add a new commit the bot runs a new suggestion round, so this has never been mentioned but came up now.

@kkovaletp What is your opinion on this? Should I change it to log an error and then return instead? That would display an error in the console (hidden unless people know to open the console and look)

googollee reviewed (approved):
LGTM. This PR takes a long time. I'd say we can merge it first and see how it works widely.

kkovaletp commented:
Ok, as @googollee approved it and suggested to merge it, and I have no critical objections, I'll merge this PR.
